Add Go prompt

seanhagen opened the pull request:
Adds the version of [Go](https://go.dev/) as reported by `go version` as a prompt segment:
![10-06-2022 15:27:26](https://user-images.githubusercontent.com/444767/173158515-da09890d-eb30-4e10-af74-f88ad8d05800.png)

bobthecow commented on functions/fish_prompt.fish:
```diff
@@ -803,6 +804,22 @@ function __bobthefish_rvm_info -S -d 'Current Ruby information from RVM'
+    
+    set -l go_version
+    if type -fq go
+        set go_version (go version | cut --delimiter=' ' -f 3 | cut --characters='3-6' )
```
first, `cut` isn't particularly portable. for BSD / macOS, this would probably have to be

```fish
go version | cut -d ' ' -f 3 | cut -c '3-6'
```

but! we can probably do this with a single builtin rather than piping to `cut` (twice!) maybe this?

```fish
go version | string match -r 'go version go(\\d+\\.\\d+)' -g
```

seanhagen replied:
I didn't know `cut` wasn't part of the basic set of CLI utilities on BSD or macOS systems! The `string match` version works for me so I'll switch it to that.

bobthecow replied:
Cut does exist, but the long option names don't work on the BSD flavor (which macOS uses). 

Regardless, using a Fish builtin wherever possible saves us processing time, so we like to keep exceptions to a minimum 🙂

bobthecow commented on functions/fish_prompt.fish:
```diff
@@ -1172,6 +1189,7 @@ function fish_prompt -d 'bobthefish, a fish theme optimized for awesome'
     __bobthefish_prompt_rubies
+    __bobthefish_prompt_golang
     __bobthefish_prompt_virtualfish
     __bobthefish_prompt_virtualgo
```
should this be combined with virtualgo? e.g. the `rubies` segment uses whatever is most relevant of the many ways to define a ruby version.

seanhagen replied:
I'd be down with that; I'll probably also see if I can steal the "only show Ruby version when in a folder where it matters" bit from the `rubies` segment too ( more thoughts on this in [my other comment](https://github.com/oh-my-fish/theme-bobthefish/pull/327#issuecomment-1155525636) )

bobthecow reviewed (commented):
One of the principles of bobthefish is "only show things that are interesting". This is why, for example, it omits default branch names in git. In ruby, it only shows the local ruby version if it's not the system ruby version. Is there an equivalent here? Should we ever suppress the go version?

I imagine if (like me) you have system go installed, but never actually do anything with it ever, it would just be noise to have it show up in every prompt. If that's as common a scenario as I think, we should probably either (1) not show if go version == system go, or (2) make it opt-in rather than opt-out.

Thoughts?

seanhagen commented:
Actually, just realized there's a better way to do this based on what you said.

Instead of always showing the version based on the `go version` command, only show the version if you're in a folder that has a `go.mod` file ( or there is a `go.mod` in a folder above the one you're in ). And instead of showing the output of `go version` show the version from `go.mod`?

And you're right about it being noise, I'm already considering turning it off and I've only had it for a few days now. I think what's more useful is showing the version of Go only if you're in a Go project ( folder with `go.mod`, or a child folder inside that project ); and changing the color if the output of `go version` is lower than what's in the `go.mod`.

That way, you can see immediately if the version of Go you have is the right one when you check out a project ( or if the version of Go you have installed/configured with a version manager like [asdf](https://asdf-vm.com/) isn't the right one for the project ).

Also, maybe a flag so that it only shows if the version of Go you have isn't high enough for the project that you're in; kind of the way the Git stuff only shows if there's changes?

bobthecow commented:
Yep, that sounds consistent, and like it'd be a much better experience :)

bobthecow commented on functions/fish_prompt.fish:
```diff
@@ -846,6 +847,85 @@ function __bobthefish_prompt_rubies -S -d 'Display current Ruby information'
+
+    set -l actual_go_version "0"
+    if ! type -fq go
+        set no_go_installed 1
```
what are we doing with this variable?

bobthecow commented on functions/fish_prompt.fish:
```diff
@@ -846,6 +847,85 @@ function __bobthefish_prompt_rubies -S -d 'Display current Ruby information'
+        set -l asdf_golang_version (asdf current golang 2>/dev/null)
+        or set -l asdf_golang_version "na"
+        
+        set -l asdf_go_sdk_version (asdf current go-sdk 2>/dev/null)
```
if we already got a value above, there's no need to check here, right? save ourselves the `asdf` call?

it sacrifices a tiny bit of readability, but saves a decent amount of complexity to inline all the things :) 

```fish
if type -fq asdf
  asdf current golang 2>/dev/null | read -l __ asdf_go_version __
  or asdf current go-sdk 2>/dev/null | read -l __ asdf_go_version __
  and set go_version $asdf_go_version
end
```

bobthecow replied:
oh, sorry, i didn't realize you were still changing things :)

bobthecow commented on functions/fish_prompt.fish:
```diff
@@ -846,6 +847,85 @@ function __bobthefish_prompt_rubies -S -d 'Display current Ruby information'
+            break
+        end
+
+        cd $dir/..
```
we don't need to actually `cd` around the filesystem. We just want to find and read a file, right?

check out how the `hg` implementation looks for a `.hg` directory. it's almost exactly what we want to do here.

https://github.com/oh-my-fish/theme-bobthefish/blob/332f23abd7a095d5b2c024a061af7b890a4f0c20/functions/fish_prompt.fish#L230-L242

bobthecow commented on functions/fish_prompt.fish:
```diff
@@ -846,6 +847,85 @@ function __bobthefish_prompt_rubies -S -d 'Display current Ruby information'
+
+        if test -f "$gomod_file"
+            set found_gomod 1
+            cat "$gomod_file" | grep "^go\ " | read _gomod gomod_version 
```
we don't ever need to pipe `cat` to `grep` :) 

```fish
grep "^go\ " "$gomod_file" | read __ gomod_version
```

seanhagen replied:
😅

bobthecow commented on functions/fish_prompt.fish:
```diff
@@ -846,6 +847,85 @@ function __bobthefish_prompt_rubies -S -d 'Display current Ruby information'
+        set go_version $gomod_version
+    end
+
+    if test "$found_gomod" -eq "0"
```
can we use the existence of `$gomod_version` rather than setting another variable?

bobthecow commented on functions/fish_prompt.fish:
```diff
@@ -846,6 +847,85 @@ function __bobthefish_prompt_rubies -S -d 'Display current Ruby information'
+    echo -ns $go_glyph
+    if test "$high_enough_version" -eq "0"
+        # the version of go 
+        __bobthefish_start_segment $color_rvm
```
doing the glyph in one color and the rest of the segment in another feels a bit awkward (and inconsistent with, e.g. the `git` segment). let's set the whole segment to the other color?

seanhagen replied:
I was thinking that it could be a warning that the version of Go someone has doesn't match the version in the `go.mod` file, letting them know they might not be able to compile the project

seanhagen commented:
Okay, changed it so it up a bit, here's what changed.

First up, I added two variables: `theme_display_go_show_wrong_version` and `theme_display_go_actual`. More on those in a bit.

Next, check to see if we're in a Go project by looking for a `go.mod` file. If one isn't found don't display the Go version -- like how the Git prompt works.

Next, check if there's a Go executable. Not worrying about asdf or virtualgo, just if there is a binary called `go` in the PATH. If there is a Go executable, use it to grab the current version.

After that, check if the version reported by `go` meets the required version from the `go.mod`. 

What the prompt displays depends on three things:
 - the value in `theme_display_go_show_wrong_version`
 - the value in `theme_display_go_actual`
 - the version of Go from `go version` and what's in the `go.mod`

If `theme_display_go_show_wrong_version` is set to "yes" and the version reported by `go` isn't equal to or higher than the version from the `go.mod` file the version is printed out in red ( or whatever color `color_rvm` is set to ). I thought this might be useful, as it lets the user know that the version of Go they're currently using can't ( or at least _probably_ can't ) compile this project.

After that, we check the value of `theme_display_go_actual`; it can be set to "no", "yes", or "diff". If it's set to "no", the version from the `go.mod` file is displayed. If it's set to "yes", the version from `go version` is displayed. 

If it's set to "diff", it shows just the version from `go.mod` if it matches `go version`, or the `go.mod` version **followed** by the _actual_ `go version` version. For example, if the version set in `go.mod` is "1.12" and the version reported by `go version` is "1.18" it will show "1.12 (1.18)".

With those two variables the user can control:
 - if they get a warning in the prompt if the version from `go version` isn't high enough for the project
 - if they see just the version from `go.mod`, just the version from `go version`, or both ( if they're not the same )

seanhagen commented:
@bobthecow I think it's ready for another look now

bobthecow reviewed (commented):
so the things we're trying to solve for are:

1. am i in a go project?
2. what go version _should_ i have?
3. do i have the right go version?
4. what go version _do_ i have?

i think we can cover all of those with just three options:

- `off`: never display any golang segment.
- `on`: regular colored segment with required version, or error colored segment, also with required version.
- `verbose`: regular colored segment with required version, or "error" colored segment with required version _and actual version_.

i could also see just the golang segment, but no version number, for the non-error state in `on`. i don't do enough go development to know whether this is okay, or if we need to know the version number in a non-error state.

the basic principle is only putting high signal things in the prompt. enumerate all the cases we care about, only show interesting things (e.g. when our expectations are not met) and letting the lack of a thing be one of the states.

seanhagen commented:
>
>  i think we can cover all of those with just three options:
> 
>     * `off`: never display any golang segment.
>     * `on`: regular colored segment with required version, or error colored segment, also with required version.
>     * `verbose`: regular colored segment with required version, or "error" colored segment with required version _and actual version_.
> 
> i could also see just the golang segment, but no version number, for the non-error state in `on`. i don't do enough go development to know whether this is okay, or if we need to know the version number in a non-error state.
> 

I went with keeping the version number always visible, even in the "ok" state because it's a piece of information that's handy to surface -- probably for the same reasons the Ruby segment does

seanhagen commented:
Lemme know if there is anything else that needs tweaking or polishing!

fabiomargarido commented:
Hey there.
It's been almost 2 years since this has seen any activity, are there still any plans to include these changes? I personally think this is helpful and have been using it (locally modifying my git copy) for quite some time.

bobthecow commented:
you're right. this should have been merged a long time ago 😞

fabiomargarido commented:
no problems, thanks for the quick response and for integrating it 🎉
